%%HTML
<h3 align="center">Министерство науки и высшего образования Российской Федерации</h3>
<h3 align="center"> Федеральное государственное бюджетное образовательное учреждение высшего образования </h3>
<h3 align="center"> Уфимский государственный нефтяной технический университет </h3>
<h3 align="center"> “Высшая школа информационных и социальных технологий” </h3>
<h3 align="center"> Кафедра “Информационные технологии и прикладная математика” </h3>
<br>
<br>
<br>
<h3 align="center"> Отчет </h3>
<h3 align="center"> по лабораторной работе №4 </h3>
<br>
<br>
<br>
<h3 align="right"> Выполнил: </h3>
<h3 align="right"> Ст-гр: БНИ-21-01 Салиев Альберт Русланович </h3>
<h3 align="right"> Проверил: </h3>
<h3 align="right"> Гумерова Велена Ильгизовна </h3>
<h3 align="right"> Дата: 25.04.2024 </h3>
<br>
<br>
<br>
<h3 align="center"> Уфа 2024 </h3>
<br>
<br>

%%HTML
<h3 align="left"> Подключены необходимые библиотеки </h3>

</h3>

%%HTML
<h3 align="left"> Подготовлены данные для кластеризации </h3>

</h3>

In [2]:
D <- read.csv("BankChurners.csv")
D <- as.data.frame(D)

In [3]:
D$age_range <- cut(D$Customer_Age, breaks = c(0, 41, 52, Inf), labels = c("Under 41", "41-52", "Over 52"))

D$Churn=as.factor(if_else(D$Attrition_Flag=="Attrited Customer",1,0))

%%HTML
<h3 align="left"> Выбрали атрибуты которые наиболее влияют на Attrition_Flag </h3>

</h3>

In [4]:
nu<-c("Total_Relationship_Count","Months_Inactive_12_mon","Contacts_Count_12_mon","Credit_Limit"
  ,"Total_Revolving_Bal","Total_Amt_Chng_Q4_Q1","Total_Trans_Amt","Total_Trans_Ct","Total_Ct_Chng_Q4_Q1"
  ,"Avg_Utilization_Ratio","Gender","Income_Category","Churn")

df<-D[,nu]
dim(df)

[1] 10127    13

In [5]:
table(df$Churn)


   0    1 
8500 1627 

%%HTML
<h3 align="left"> Выполнено разделение на выборки для обучения и проверки в соотношении 8:2 </h3>

</h3>

In [6]:
set.seed(123)
trainIndex <- createDataPartition(df$Churn, p = .8, list = FALSE)
training <- df[trainIndex,]
testing <- df[-trainIndex,]

In [7]:
table(training$Churn)


   0    1 
6800 1302 

%%HTML
<h3 align="left"> Расширим тренировочный датасет </h3>

</h3>

In [8]:
training <- upSample(x = subset(training, select = -c(Churn)),
                     y = training$Churn,yname = "Churn") 

table(training$Churn)


   0    1 
6800 6800 

%%HTML
<h3 align="left"> Проведем логистическую регрессию и проверим точность модели </h3>

</h3>

In [9]:
## Logistic Regression
model1 = glm(formula=Churn ~ . , family = binomial(link='logit'),data = training) %>% stepAIC()
summary(model1)

probabilities = predict(model1,newdata=testing,type='response')

prob <- ifelse(probabilities >= 0.5, 1,0)
table(prob)

Start:  AIC=9414.12
Churn ~ Total_Relationship_Count + Months_Inactive_12_mon + Contacts_Count_12_mon + 
    Credit_Limit + Total_Revolving_Bal + Total_Amt_Chng_Q4_Q1 + 
    Total_Trans_Amt + Total_Trans_Ct + Total_Ct_Chng_Q4_Q1 + 
    Avg_Utilization_Ratio + Gender + Income_Category

                           Df Deviance     AIC
- Credit_Limit              1   9381.6  9413.6
<none>                          9380.1  9414.1
- Avg_Utilization_Ratio     1   9384.9  9416.9
- Income_Category           5   9415.7  9439.7
- Total_Amt_Chng_Q4_Q1      1   9425.0  9457.0
- Gender                    1   9496.8  9528.8
- Total_Revolving_Bal       1   9724.2  9756.2
- Months_Inactive_12_mon    1   9733.7  9765.7
- Contacts_Count_12_mon     1   9793.9  9825.9
- Total_Ct_Chng_Q4_Q1       1   9863.8  9895.8
- Total_Relationship_Count  1   9887.4  9919.4
- Total_Trans_Amt           1  10607.2 10639.2
- Total_Trans_Ct            1  13060.2 13092.2

Step:  AIC=9413.64
Churn ~ Total_Relationship_Count + M


Call:
glm(formula = Churn ~ Total_Relationship_Count + Months_Inactive_12_mon + 
    Contacts_Count_12_mon + Total_Revolving_Bal + Total_Amt_Chng_Q4_Q1 + 
    Total_Trans_Amt + Total_Trans_Ct + Total_Ct_Chng_Q4_Q1 + 
    Avg_Utilization_Ratio + Gender + Income_Category, family = binomial(link = "logit"), 
    data = training)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-3.3310  -0.4615   0.0092   0.4984   2.9281  

Coefficients:
                                Estimate Std. Error z value Pr(>|z|)    
(Intercept)                    8.332e+00  2.424e-01  34.370  < 2e-16 ***
Total_Relationship_Count      -3.946e-01  1.822e-02 -21.661  < 2e-16 ***
Months_Inactive_12_mon         5.127e-01  2.799e-02  18.315  < 2e-16 ***
Contacts_Count_12_mon          4.908e-01  2.513e-02  19.533  < 2e-16 ***
Total_Revolving_Bal           -7.950e-04  4.174e-05 -19.046  < 2e-16 ***
Total_Amt_Chng_Q4_Q1          -8.742e-01  1.325e-01  -6.599 4.15e-11 ***
Total_Trans_Amt                

prob
   0    1 
1486  539 

%%HTML
<h3 align="left"> Также испольуя confusion matrix </h3>

</h3>

In [10]:
tab <- table(actual= testing$Churn,predicted=prob)
caret::confusionMatrix(tab,positive = "1")

Confusion Matrix and Statistics

      predicted
actual    0    1
     0 1433  267
     1   53  272
                                          
               Accuracy : 0.842           
                 95% CI : (0.8253, 0.8576)
    No Information Rate : 0.7338          
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.5369          
                                          
 Mcnemar's Test P-Value : < 2.2e-16       
                                          
            Sensitivity : 0.5046          
            Specificity : 0.9643          
         Pos Pred Value : 0.8369          
         Neg Pred Value : 0.8429          
             Prevalence : 0.2662          
         Detection Rate : 0.1343          
   Detection Prevalence : 0.1605          
      Balanced Accuracy : 0.7345          
                                          
       'Positive' Class : 1               
                                        

%%HTML
<h3 align="left"> Наблюдается точность 84,2% с надежностью 95% в интервале (82.53% , 85.67%) </h3>

<h3 align="left"> Далее проведена тренировка модели на основе Наивного байесовского классификатора  </h3>

<h3 align="left"> И проверена её  точность  </h3>

</h3>

In [11]:
##Naive Bayes classifier
classifier_cl <- naiveBayes(Churn ~ ., data =training)
# Predicting on train data'
y_pred <- predict(classifier_cl, newdata = testing)
# Confusion Matrix
cm <- table(actual=testing$Churn,predicted= y_pred)
caret::confusionMatrix(testing$Churn,y_pred,positive = "1")

Confusion Matrix and Statistics

          Reference
Prediction    0    1
         0 1324  376
         1   59  266
                                          
               Accuracy : 0.7852          
                 95% CI : (0.7666, 0.8029)
    No Information Rate : 0.683           
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.4283          
                                          
 Mcnemar's Test P-Value : < 2.2e-16       
                                          
            Sensitivity : 0.4143          
            Specificity : 0.9573          
         Pos Pred Value : 0.8185          
         Neg Pred Value : 0.7788          
             Prevalence : 0.3170          
         Detection Rate : 0.1314          
   Detection Prevalence : 0.1605          
      Balanced Accuracy : 0.6858          
                                          
       'Positive' Class : 1               
                        

%%HTML
<h3 align="left"> Наблюдается точность 78,5% с надежностью 95% в интервале (0.7666, 0.8029) </h3>

<h3 align="left"> Далее проведена тренировка модели с помощью метода опорных векторов(SVM) </h3>

</h3>

In [12]:
# SVM Linear
svmL.fit = svm(Churn ~ ., data = training, kernel = "linear")

svmL_predication = predict(svmL.fit, testing)

svm <- table(actual=testing$Churn,predicted= y_pred)
caret::confusionMatrix(testing$Churn,svmL_predication,positive = "1")


Confusion Matrix and Statistics

          Reference
Prediction    0    1
         0 1428  272
         1   52  273
                                          
               Accuracy : 0.84            
                 95% CI : (0.8233, 0.8557)
    No Information Rate : 0.7309          
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.5339          
                                          
 Mcnemar's Test P-Value : < 2.2e-16       
                                          
            Sensitivity : 0.5009          
            Specificity : 0.9649          
         Pos Pred Value : 0.8400          
         Neg Pred Value : 0.8400          
             Prevalence : 0.2691          
         Detection Rate : 0.1348          
   Detection Prevalence : 0.1605          
      Balanced Accuracy : 0.7329          
                                          
       'Positive' Class : 1               
                        

%%HTML
<h3 align="left"> Наблюдается точность 84% с надежностью 95% в интервале (0.8233, 0.8557) </h3>

<h3 align="left"> Видим, что наилудшую точность имеет логистическая регрессия</h3>

</h3>